Semantics.md

```diff
@@ -683,3 +683,22 @@ outside the range which rounds to an integer in range) traps.
 [future types]: FutureFeatures.md#more-table-operators-and-types
 [future large pages]: FutureFeatures.md#large-page-support
 [future ieee 754]: FutureFeatures.md#full-ieee-754-2008-conformance
+
+
+## Validation
+
+A module binary has to be _validated_ before it is compiled.
+Validation ensures that the module is well-defined and that its code cannot exhibit any undefined behavior.
+In particular, along with some runtime checks, it ensures that no program can access or corrupt memory it does not own.
+
+Validation of code is mostly defined in terms of type-checking the use of the operand stack.
+It sequentially checks for each that the expected operands can be popped from the stack, and tracks which new operands are pushed onto it. At the start of a function the stack is empty; at its end it has to match the return type of the function. In addition, instructions inside a `block` (or `loop` or `if`) cannot consume operands pushed outside; at the end of the block the remaining inner operands must match the block signature.
+
+A special case are instructions performing an unconditional control transfer (`br`, `br_table`, `return`, `unreachable`), because execution never proceeds after them.
+The stack after such an instruction is unconstrained, und thus said to be _polymorphic_.
+Instructions after an unconditional control transfer still need to type-check,
+but conceptually, values of any type can be popped off a polymorphic stack for the sake of checking consecutive instructions.
+A polymophic stack also matches any possible signature at the end of a block or function.
```

Comment:
@Cellule That still does not read well, how can "The stack after such an instruction is unconstrained" yet values of specific types can be pushed onto the stack and must type check? 

Also what does "A polymophic stack also matches any possible signature at the end of a block or function." if the code after the unreachable operator is to be typed checked, are the values discarded and ignored, or must the stack have just the expected number of values?

There is the following in the spec, and this PR just references this, but honestly I find this cryptic and it is impossible to evaluate the merits of the PR in a reasonable time and it need not be take cryptic. Looking at the `push` does it fail if pushing a value to a stack with `Ellipses`, and if so that would break the checking after the unreachable operator, or am I just reading this wrong (highly likely but a good description would help).  The `Ellipses` stack property also appears sticky, so it's not clear it is doing what you expect in your examples. Sorry I don't even follow what the `pop` operation is doing, a description would help.

```
let pop (ell1, ts1) (ell2, ts2) at =
  let n1 = List.length ts1 in
  let n2 = List.length ts2 in
  let n = min n1 n2 in
  let n3 = if ell2 = Ellipses then (n1 - n) else 0 in
  check_stack ts1 (Lib.List.make n3 None @ Lib.List.drop (n2 - n) ts2) at;
  (ell2, if ell1 = Ellipses then [] else Lib.List.take (n2 - n) ts2)

let push (ell1, ts1) (ell2, ts2) =
  assert (ell1 = NoEllipses || ts2 = []);
  (if ell1 = Ellipses || ell2 = Ellipses then Ellipses else NoEllipses),
  ts2 @ ts1
```